[Utils] Add snapcraft support

cevap opened the pull request:
To be done before merging this pull:

- [ ] PIVX should register own snapcraft and adapt links, I used PIVX-Project/PIVX for this pull
  - [ ] Before you [register](https://snapcraft.io/account), it would be probably better if builds built from PIVX-Project are called pivx and not some other name.
    
    _Pivx is reserved for [my master branch](https://github.com/cevap/PIVX/) and if you merge this pull, then I will rename pivx on cevap to pivx-cevap (or whatever you might define for community/contribution)_.
    _You can login to [snapcraft](https://snapcraft.io/login?next=/account) via your github account but to manage builds and releases you need to have [launchpad account](https://login.ubuntu.com/)_
- [ ] Add text/information about PIVX, [dummy one which I created](https://snapcraft.io/pivx/) is:
```
PIVX is an MIT licensed,
open source, blockchain-based cryptocurrency with 
ultra fast transactions, low fees, high network decentralization, and
Zero Knowledge cryptography proofs for industry-leading
transaction anonymity. 

Download PIVX White Paper PDF:
  https://pivx.org/wp-content/uploads/2018/10/PIVX-White.pdf
```

- [ ] Optional to be done after merging: trigger rebuilds for older releases
    _PIVX team should probably define from which branch you want to release, if you use 4 channels or not (2 of them are for stable releases: stable, candidate, 2 of them for development releases: beta and edge)_
    
    this is how I could imagine it to work 

branch                              | tag                           | channel   | when built/released 
:------------------------------- | :------------------------ | -----------: | :-------------------------------------------------
master                              | ***no tag***              | edge        | built & released: each commit to master
***not built from branch*** | PreRelease             | beta         | built & released by retriggering snap with correct tag/branch
***not built from branch*** | Release Candidate | candidate | built & released by retriggering snap with correct tag/branch
***not built from branch*** | Final release           | stable       | built & released by retriggering snap with correct tag/branch. ***Release to stable channel should be done only manually***

Snap documentation: https://docs.snapcraft.io/

### Example of use with troubleshooting
Screenshot of current (at the time of creation of this pull) builds, all are built successfully, i386 did not release automaticly:
![grafik](https://user-images.githubusercontent.com/24996551/50355579-7da48500-054f-11e9-8b82-4f90f678f8f1.png)

In such case you can easily release them from [your dashboard](https://snapcraft.io/pivx/releases), switch to recent releases and if any of them was not released (for whatever reason/s), you can do it manually:
![grafik](https://user-images.githubusercontent.com/24996551/50355792-1c30e600-0550-11e9-8597-88249cbfb1b7.png)


![grafik](https://user-images.githubusercontent.com/24996551/50355764-09b6ac80-0550-11e9-8315-414e3f05154c.png)

![grafik](https://user-images.githubusercontent.com/24996551/50355823-3a96e180-0550-11e9-98ed-de47520ea4f4.png)

then click on promote revisions:
![grafik](https://user-images.githubusercontent.com/24996551/50355861-4b475780-0550-11e9-886d-cd94a5be0be2.png)

You can see that last build is now available over edge channel, to promote new channel, simply click on them source channel you want to promote from (in current case it is edge to beta):
![grafik](https://user-images.githubusercontent.com/24996551/50355994-bc870a80-0550-11e9-9908-a96f5a3a497d.png)

Apply changes and it it is published:
![grafik](https://user-images.githubusercontent.com/24996551/50356025-d9234280-0550-11e9-8cce-c42237cc005c.png)

cevap commented:
testing travis on my own branch https://travis-ci.org/cevap/PIVX/builds/472290833, will revert README.md

this pull includes some non required commits, I can make clean pull but you can edit too.

Mrs-X commented:
Is this completed/finished?

How can it be tested?

Fuzzbawls commented:
The `snap/` directory (and it's child files) should probably be placed under the existing `contrib/` directory rather than the repository's root if at all possible. That would be more in-line with where existing 3rd party supporting files are located.

```
pivx/
└── contrib/
    └── snap/
        ├── gui/
        │   ├── pivx.desktop
        │   └── pivx128.png 
        ├── patches/
        │   └── default_data_dir.patch
        └── snapcraft.yaml
...
```
Also, you've got some trailing white space in `snapcraft.yaml` which obviously doesn't affect any build tests, but causes Travis to complain and error out.

cevap commented:
> The `snap/` directory (and it's child files) should probably be placed under the existing `contrib/` directory rather than the repository's root if at all possible. That would be more in-line with where existing 3rd party supporting files are located.
> 
> ```
> pivx/
> └── contrib/
>     └── snap/
>         ├── gui/
>         │   ├── pivx.desktop
>         │   └── pivx128.png 
>         ├── patches/
>         │   └── default_data_dir.patch
>         └── snapcraft.yaml
> ...
> ```

I am sure it can be placed anywhere (_even outside, to another repo_) and adapting scripts make it possible.

There are features and options provided, if PIVX defines that you do not want automatic building etc (_actually everything which makes snap quite nice_), then you can move it and use scripting. Snapcraft looks for snapcraft.yaml in root and in snap folder. if moving from there, then it will break http://snapcraft.io/build process. That is the main reason why I put it to snap folder, as it is the default one for snap builds.

For the coin and the patch, again, this is just showcase how to use and create them, there is still no testnet icon available as I am even not aware if you keep same icon for snap builds (_maybe you want a different one for ability of visual differentiation for end users_)
> 
> Also, you've got some trailing white space in `snapcraft.yaml` which obviously doesn't affect any build tests, but causes Travis to complain and error out.

I changed descriptions, should be fixed with this commit https://github.com/PIVX-Project/PIVX/pull/782/commits/23c74029ece2ad81c4d26599d81fa644583829dc

Travis fails with this error:
```
subtree commit 452d8e4d2a2f9f1b5be6b02e18f1ba102e5ca0b4 unavailable: cannot compare
```

If you register snapcraft to suggested names, let me know to which user email I should transfer ownership of pivx snap name.

cevap commented:
> 
> Also, you've got some trailing white space in `snapcraft.yaml` which obviously doesn't affect any build tests, but causes Travis to complain and error out.

I changed descriptions, should be fixed now.

cevap commented:
Added a break for snapcraft description
```
  
  Download PIVX White Paper PDF:
    https://pivx.org/wp-content/uploads/2018/10/PIVX-White.pdf
```

without `  ` in description as break line, snap would break, I will delete that break (https://github.com/PIVX-Project/PIVX/pull/782/commits/94e3c63d92a25ef32419279d595062fd6d6a40b7) then travis will work like [here before I added a break](https://travis-ci.org/PIVX-Project/PIVX/builds/482166150?utm_source=github_status&utm_medium=notification)

cevap commented:
@Fuzzbawls 
you can run snapcraft build on your local pc and test it, here is a message which you can see:
```
The snap/ directory is meant specifically for snapcraft, but it contains the following non-snapcraft-related paths, which is unsupported and will cause unexpected behavior:
- gui/ion-qt.ico
- patches
- patches/default_data_dir.patch

If you must store these files within the snap/ directory, move them to snap/local/, which is ignored by snapcraft.
```

I just assume you did not build/try it. To build locally, [follow these steps](https://docs.snapcraft.io/build-on-lxd/4157). On ubuntu 16.04 and above, you don't have to install anything, everything is preinstalled and to build:

1. Set build environment (I use here building without bases, check guide above for other possibilities)
    Set `SNAPCRAFT_BUILD_ENVIRONMENT` or add it to your `.profile`
    `echo "export SNAPCRAFT_BUILD_ENVIRONMENT=lxd" >> ~/.profile`
2. cd into PIVX folder which includes snap/snapcraft.yaml file (or copy it manually)
3. run `snapcraft cleanbuild` or `snapcraft cleanbuild --debug` if you want to debug.

If you need to rebuild everything and to cleanup, you can run `snapcraft clean`, for more info run `snapcraft --help`, you can release and deploy per command line too, please check snapcraft documentation for more info.

cevap commented:
I replaced building with shared libs, now building for different platforms works and builds with depends folder. There were several caveats to get this work, here is a list of main troubles for a reference later:

1. Wallet becomes corrupt when it gets bigger
    By using berkeleydb, it is compiled to use /var/tmp folder, which is not accessable with [strict confinment](https://docs.snapcraft.io/snap-confinement/6233) on snap. One solution would be to change confinment mode to classic or dev mode, but this would make the snap non installable on ubuntu core, which is bad. Beside that, strict confinment provides much more secure environment.
  my solution: I added a [patch into snap folder](https://github.com/ioncoincore/ion/blob/master/snap/patches/fix-bdb-tmp-folder.patch) which [does patch bdb.mk](https://github.com/ioncoincore/ion/blob/a9d7aeac5a3da4faffab0b20fe2176a6a6146d67/snap/snapcraft.yaml#L87) to change used folder to the one in snap folder which is writeable, with that we keep strict confinment mode and berkeleydb works as it should.
1. User has no access to his HOME folder. On one side this is great, as it ensures me that it is secure and it will not harm any files in my home folder. On other side, users do want to use their HOME/.pivx folder. To resolve access to HOME for write access, I added plug HOME to the snapcraft which does allow the app access to the HOME folder. Additionaly, you can open Ubuntu Software Center, choose pivx, click on permissions and there is a slider to switch on/off access to HOME folder, per default it is on.
1. Same like with home is folder mount folder (/mnt). If somebody wants to backup wallets on mounted USB/HDD, he will not be able to access it with strict confinment. For access, I added according plug and now mount folder and everything inside has read/write access.

Additional info:
If you build from snapcraft, not all platforms are available. You should also register on launchpad and create there snap package, that will build snap also for powerppc architecture.
![image](https://user-images.githubusercontent.com/24996551/52295715-ca67e200-297c-11e9-8b45-ca2c823b2602.png)

Launchpad can also build automaticly:
![image](https://user-images.githubusercontent.com/24996551/52295758-e075a280-297c-11e9-8ebd-4782561236e2.png)

I will update snap part for PIVX and push it in next 24 hours.

If you want other changes in snap, please let me know.

There is quite a lot additional info why and how cool snap is. Yes, you can also run snap builds on mac. Yes, you can run snap builds on windows. Please read https://docs.snapcraft.io/ for further documentation

@Mrs-X 
Actually it is finished, I waited for PIVX to create snapcraft account so that I can transfer reserved name "pivx" to this team and then apply changes and trigger/retrigger builds.

![image](https://user-images.githubusercontent.com/24996551/52296021-9f31c280-297d-11e9-8187-e31e242036a2.png)

cevap commented:
@Mrs-X how to test:

- install stable: `sudo snap install pivx` (installs default stable channel)
- Install latest edge: `sudo snap install --edge pivx`
![image](https://user-images.githubusercontent.com/24996551/52296442-ae654000-297e-11e9-9c44-8a6e2168ae80.png)

If you install it now, it will install last PIVX from my pivx forked repo. You actually can verify first 3 points of my last post with that build, as BDB will fail to work properly with currently available snap online. Like I said in previous post, I will update today snapcraft for PIVX and it will be built with depends folder, not shared libs installed, this is actually perfectly working snap in secure environment (strict confinment) which I will apply on PIVX: https://github.com/ioncoincore/ion/blob/master/snap/snapcraft.yaml

cevap commented:
created a test build locally, download: https://github.com/cevap/PIVX/releases/tag/v3.1.99-snap-x86_64

cevap commented:
this snapcraft works also with launchpad if you create snap on your launchpad. Launchpad would provide you additional architecture: powerpc

cevap commented:
current snap built for all platforms, includes regtest fix
who wants to test it:
```bash
sudo snap install pivx
```

Commands to start daemon, QT, cli
```bash
pivx.qt
pivx.daemon
pivx.cli
```
![grafik](https://user-images.githubusercontent.com/24996551/52543177-c6d2c180-2da7-11e9-89a1-3f441a46ffd7.png)

![grafik](https://user-images.githubusercontent.com/24996551/52543187-d225ed00-2da7-11e9-8259-3a40b549b86a.png)

![grafik](https://user-images.githubusercontent.com/24996551/52543190-db16be80-2da7-11e9-95f6-9f1f79959ec8.png)

![grafik](https://user-images.githubusercontent.com/24996551/52543198-e964da80-2da7-11e9-9f3b-af35bfa0d27a.png)

Fuzzbawls commented:
@cevap revisiting this a bit now that some more time is available to me.

I am able to create a `.snap` file for the host arch (amd64) no problem, but attempting to build for other archs is proving to be problematic

Fuzzbawls commented:
```
Setting target machine to 'aarch64'
Downloading parts list|                                                                                                                                                                                            
Sorry, an error occurred in Snapcraft:
Traceback (most recent call last):
  File "/snap/snapcraft/2722/legacy_snapcraft/bin/snapcraft", line 11, in <module>
    load_entry_point('snapcraft==2.43.1', 'console_scripts', 'snapcraft')()
  File "/snap/snapcraft/2722/legacy_snapcraft/lib/python3.5/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/snap/snapcraft/2722/legacy_snapcraft/lib/python3.5/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/snap/snapcraft/2722/legacy_snapcraft/lib/python3.5/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/snap/snapcraft/2722/legacy_snapcraft/lib/python3.5/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/snap/snapcraft/2722/legacy_snapcraft/lib/python3.5/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/snap/snapcraft/2722/legacy_snapcraft/usr/lib/python3.5/contextlib.py", line 77, in __exit__
    self.gen.throw(type, value, traceback)
  File "/snap/snapcraft/2722/legacy_snapcraft/lib/python3.5/site-packages/click/core.py", line 87, in augment_usage_errors
    yield
  File "/snap/snapcraft/2722/legacy_snapcraft/lib/python3.5/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/snap/snapcraft/2722/legacy_snapcraft/lib/python3.5/site-packages/snapcraft/cli/lifecycle.py", line 187, in prime
    _execute(steps.PRIME, parts, **kwargs)
  File "/snap/snapcraft/2722/legacy_snapcraft/lib/python3.5/site-packages/snapcraft/cli/lifecycle.py", line 97, in _execute
    project_config = project_loader.load_config(project)
  File "/snap/snapcraft/2722/legacy_snapcraft/lib/python3.5/site-packages/snapcraft/internal/project_loader/__init__.py", line 36, in load_config
    return Config(project)
  File "/snap/snapcraft/2722/legacy_snapcraft/lib/python3.5/site-packages/snapcraft/internal/project_loader/_config.py", line 237, in __init__
    build_tools=self.build_tools,
  File "/snap/snapcraft/2722/legacy_snapcraft/lib/python3.5/site-packages/snapcraft/internal/project_loader/_parts_config.py", line 54, in __init__
    self._process_parts()
  File "/snap/snapcraft/2722/legacy_snapcraft/lib/python3.5/site-packages/snapcraft/internal/project_loader/_parts_config.py", line 85, in _process_parts
    self.load_part(part_name, plugin_name, properties)
  File "/snap/snapcraft/2722/legacy_snapcraft/lib/python3.5/site-packages/snapcraft/internal/project_loader/_parts_config.py", line 205, in load_part
    definitions_schema=self._validator.definitions_schema,
  File "/snap/snapcraft/2722/legacy_snapcraft/lib/python3.5/site-packages/snapcraft/internal/pluginhandler/_plugin_loader.py", line 79, in load_plugin
    plugin.enable_cross_compilation()
  File "/snap/snapcraft/2722/legacy_snapcraft/lib/python3.5/site-packages/snapcraft/_baseplugin.py", line 188, in enable_cross_compilation
    "to a different target architecture".format(self.name)
NotImplementedError: The plugin used by 'desktop-qt5' does not support cross-compiling to a different target architecture
```

cevap commented:
thanks for the info @Fuzzbawls , I have reworked snap, there were some additional issues beside troubles with arm/aarch due to core16 using strictly gcc-/g++-5. I upgraded it to core18 will apply new changes in next 24 hours for PIVX (have to update my master branch), but you could check latest version [here](https://github.com/ioncoincore/ion/blob/master/snap/snapcraft.yaml), as if you want to test it, just copy and paste it, here are most notable changes:

- biggest improvement is probably that due to core18 upgrade gcc7/8 (g++7/8) can be used which resolves issues with arm/aarch64 builds, because core16 does use gcc5/g++5 for compilation and there is a bug on arch, thats actually major reason which forced me to upgrade to core18 
- upgraded to core18, it builds on launchpad and is quite cool feature to build daily deb packages as well as snap directly from launchpad.
  - snapcraft does also support core18
- removed desktop-launch
- patch icons during compilation to be somehow visually different from those installed as deb/custom compilations, to our project, I added overlay snapbuild on all icons
- add desktop icon for daemon
- add temp workaround for ppc64el not building qt which is defined in apps, as workaround I create dummy sh for test and test qt bins
- patch client version with a suffix `-snap`
- make script compatible with other projects, even if their binaries use different naming schemes

Upgrade to core18 might resolve your issue, I am not sure what it is, would need to reproduce it to see, can I have more info about OS and device you compiled it on? Did you compile on same arch or as example on x86_64?

I guess you could update your snapcraft.yaml from the link above instead to wait for me. Shortcuts which you do require as to be installed (better said all files that you need snap to register), they should be in snap folder, I split them into desktop and local where patches are. Most things are patched during compilation, so, actually all files except icons could be in any other folder (_because I was asked if snapcraft.yaml can be moved to another folder_).

What I also see is if you go to https://snapcraft.io/store you will see PIVX as suggested/top app.
![grafik](https://user-images.githubusercontent.com/24996551/56476490-e3bafd80-6498-11e9-8da3-e406f4b7effc.png)

You already have launchpad, there you need to import PIVX repo: https://code.launchpad.net/+code-imports/+new and keep git system, then create SNAP from imported git branch which lets launchpad scan it each 5-6 hours before it imports again pivx master (or whatever you have choosen). That is the way how I build deb and snap packages from same commit. As I wrote in first comment, I would give the ownership of pivx name to the official account which will release pivx, so that you guys can manage it and have full control.

cevap commented:
one of travis tests fails with last commit (has nothing to do with snapcraft changes, I made PR from pivx master to current PR's branch, maybe something went wrong on that), [here is the link to travis job](https://travis-ci.org/PIVX-Project/PIVX/builds/522868376?utm_source=github_status&utm_medium=notification) and here is the error:
```shell
unknown location(0): fatal error: in "libzerocoin/libzerocoin_tests": std::runtime_error: Unable to generate random prime (too many tests)
test/libzerocoin_tests.cpp(488): last checkpoint: "libzerocoin_tests" entry.
test/libzerocoin_tests.cpp(488): Leaving test case "libzerocoin_tests"; testing time: 1573ms
test/libzerocoin_tests.cpp(487): Leaving test suite "libzerocoin"; testing time: 1573ms
Leaving test module "Pivx Test Suite"; testing time: 1573ms
*** 1 failure is detected in the test module "Pivx Test Suite"
Makefile:9927: recipe for target 'test/libzerocoin_tests.cpp.test' failed
make[3]: *** [test/libzerocoin_tests.cpp.test] Error 1
make[3]: *** Waiting for unfinished jobs....
make[3]: Leaving directory '/home/travis/build/PIVX-Project/PIVX/build/pivx-x86_64-w64-mingw32/src'
make[2]: *** [check-am] Error 2
Makefile:9683: recipe for target 'check-am' failed
make[2]: Leaving directory '/home/travis/build/PIVX-Project/PIVX/build/pivx-x86_64-w64-mingw32/src'
Makefile:9371: recipe for target 'check-recursive' failed
make[1]: *** [check-recursive] Error 1
make[1]: Leaving directory '/home/travis/build/PIVX-Project/PIVX/build/pivx-x86_64-w64-mingw32/src'
Makefile:748: recipe for target 'check-recursive' failed
make: *** [check-recursive] Error 1
```

cevap commented:
@Fuzzbawls I am quite sure that you know how to setup and use launchpad, I post this info now for informative task for other users which might want to build or import pivx into their launchpads. If building on your machine for some reason fails, for testing, it would be much faster for you to use launchpad and set testing branch from which you want to build. Here is quick guide to how to do it

How to create snap on launchpad (quick and dirty guide)
=======================================

1. (create import request)[https://code.launchpad.net/+code-imports/+new]
1.1 choose project to associate the code import with (_normally you would use your own, there is a serch if you forgot what the name is_)
  ![Bildschirmfoto von 2019-04-22 03-26-04](https://user-images.githubusercontent.com/24996551/56478829-38b73d80-64b2-11e9-807b-a0726509dd89.png)
  ![Bildschirmfoto von 2019-04-22 03-26-09](https://user-images.githubusercontent.com/24996551/56478839-4240a580-64b2-11e9-920b-a6aa376699e7.png)

1.2 name (_preset to trunk_) (_change to main if for stable packages_)
1.3 select version control system: _**GIT**_
1.4 enter your repository (example: https://github.com/PIVX-Project/PIVX.git)
  ![Bildschirmfoto von 2019-04-22 03-26-09](https://user-images.githubusercontent.com/24996551/56478839-4240a580-64b2-11e9-920b-a6aa376699e7.png)
1.4.1 *select* _**Git**__ *as target version control system*
1.5 click on "Request import"

In current example I selected PIVX-Project (official) to be associated with current snap, after clicking request import, launchpad opens me the page of imported PIVX: https://code.launchpad.net/~ionomy/pivx-project/trunk
![Bildschirmfoto von 2019-04-22 03-27-03](https://user-images.githubusercontent.com/24996551/56478893-8469e700-64b2-11e9-8e70-ff63b759cfb4.png)


2. Click "create snap package" to create new package
  ![Bildschirmfoto von 2019-04-22 03-27-15](https://user-images.githubusercontent.com/24996551/56478897-89c73180-64b2-11e9-917f-54e539c8449f.png)
2.1 select name which of the snap (_this must be the same name which you registered_)
2.2 choose series: `Ubuntu Bionic, for Ubuntu Core 16` (current PR is upgraded to bionic, if you for some reason use older versions or need powerppc build, choose ubuntu core 16)
2.3 choose processor architectures.
    current pivx snap supports all, but on ubuntu bionic corem you can build only for: amd64, arm64, armhf, i386, ppc64el and s390x
  ![Bildschirmfoto von 2019-04-22 03-42-46](https://user-images.githubusercontent.com/24996551/56478914-9ba8d480-64b2-11e9-8672-57aee8c6dae9.png)

2.4 choose if you want sources tarball to be built, for those who require it know why they require it, other can leave it unchecked.
2.5 choose if you want to build automaticly when branch changes (it means, when your github repo is updated)
    _if you build stable versions with launchpad, you should not mark this checkbox. If you do automaticly build, then keep in mind that snapcraft does build automticly too and releases builds to the edge channel, thats is why I will set here to release to beta channel which is also devel and not stable grade_
  ![Bildschirmfoto von 2019-04-22 03-42-59](https://user-images.githubusercontent.com/24996551/56478951-c4c96500-64b2-11e9-8716-21a1d1218356.png)
2.6 choose source archive for automatic builds, I have choose ~pivx/ubuntu/pivx-unstable which is I guess official one for pivx master branch releases which fits well here
2.7 change source snap channels for automatic builds (_change only if you know what you do, otherwise leave empty_)
2.8 mark checkbox "Automatically upload to store"
2.8.1 enter registered store package name, in current case it is pivx as it is registered for me
2.8.2 Risk: mark only beta channel, edge is for snapcraft builds (on each master commit, snapcraft)  and beta is updated 6 times a day (launchpad), see notes of 2.5.
2.9 click on "create snap package"
  ![Bildschirmfoto von 2019-04-22 03-43-20](https://user-images.githubusercontent.com/24996551/56478968-d874cb80-64b2-11e9-9bda-0f8f2c3f4acb.png)

other options are optional and to be changed only if you know what you do.
![Bildschirmfoto von 2019-04-22 03-43-13](https://user-images.githubusercontent.com/24996551/56478961-d27eea80-64b2-11e9-95c0-b6ec9a0076d8.png)


Waiting few seconds, and you can see "Uploads of pivx to the store are now authorized." and pivx snap build on launchpad is now here: https://code.launchpad.net/~ionomy/+snap/pivx
![Bildschirmfoto von 2019-04-22 03-44-13](https://user-images.githubusercontent.com/24996551/56478987-f4786d00-64b2-11e9-8e20-9060190cfa5e.png)

and
![Bildschirmfoto von 2019-04-22 03-44-20](https://user-images.githubusercontent.com/24996551/56478995-fe9a6b80-64b2-11e9-8c3d-12f53d76dc85.png)

We can click on "Request builds" now to start building for configured architectures or if you want, you can go to the Source by clicking on source link: lp:~ionomy/pivx-project/trunk (which is https://code.launchpad.net/~ionomy/pivx-project/trunk) and click on import now. It takes 20-30 seconds until it is imported, and as soon as it is and code was updated, launchpad will automaticly start launchpad builds.
![Bildschirmfoto von 2019-04-22 03-47-56](https://user-images.githubusercontent.com/24996551/56478999-0b1ec400-64b3-11e9-99c2-c35221912a56.png)

As I did import it now from PIVX project and not from CEVAP, current builds would fail as there is no snapcraft file in current pivx master

cevap commented:
maybe QT fix https://github.com/cevap/PIVX/commit/e2e1e085a92c6c67cab3392f43e67546a6472c38 for other architectures should be placed in separate PR, it is required here for s390x and ppc64el snap build

cevap commented:
snap is upgraded to core18 and releases built successfully. I added icons for daemon which are different to qt's in the way that specific color is converted to alpha layer, this is how it looks like on ubuntu with gnome
![grafik](https://user-images.githubusercontent.com/24996551/56541462-05d18000-656c-11e9-916a-335efbeb6be6.png)

I commented out copying of conf example and patching of images for snapcraft, I just think you will create your own and did not want to waste time on creating additional images which will be anyway probably removed.

Current builds are on `edge` channel: `sudo snap install --edge pivx`
_s390x and ppc64el will fail if (https://github.com/PIVX-Project/PIVX/pull/810) is not merged before building (boost, openssl and qt.mk files are not updated in current PR)_

@Fuzzbawls here is arm64 log: https://build.snapcraft.io/user/cevap/PIVX/536368
I added listing of all installed files as well as all vars which probably help if they are unknown, but I would think it is better to remove them to make the log shorter. Those are as example vars at the end of the part where pivx is compiled from the linked log:
```
SNAPCRAFT_IMAGE_INFO={"build_url": "https://launchpad.net/~build.snapcraft.io/+snap/e2ce211a6aa16ef668206df4054564a8/+build/536368"}
SNAPCRAFT_PROJECT_VERSION=3.2.99
SNAPCRAFT_EXTENSIONS_DIR=/snap/snapcraft/2836/share/snapcraft/extensions
SNAPCRAFT_PROJECT_GRADE=devel
SNAP_COMMON=/var/snap/snapcraft/common
SNAP_INSTANCE_KEY=
USER=root
CXXFLAGS= -I/build/pivx/parts/pivx/install/usr/include
LD_LIBRARY_PATH=:/build/pivx/parts/pivx/install/lib:/build/pivx/parts/pivx/install/usr/lib:/build/pivx/parts/pivx/install/lib/aarch64-linux-gnu:/build/pivx/parts/pivx/install/usr/lib/aarch64-linux-gnu
SNAPCRAFT_PART_SRC=/build/pivx/parts/pivx/src
OLDPWD=/build/pivx/parts/pivx/build/depends
HOME=/root
SNAP_LIBRARY_PATH=/var/lib/snapd/lib/gl:/var/lib/snapd/lib/gl32:/var/lib/snapd/void
SNAP_USER_DATA=/root/snap/snapcraft/2836
LDFLAGS= -L/build/pivx/parts/pivx/install/lib -L/build/pivx/parts/pivx/install/usr/lib -L/build/pivx/parts/pivx/install/lib/aarch64-linux-gnu -L/build/pivx/parts/pivx/install/usr/lib/aarch64-linux-gnu
SNAPCRAFT_PARALLEL_BUILD_COUNT=4
container=lxc
SNAP_REVISION=2836
SNAPCRAFT_STAGE=/build/pivx/stage
SNAPCRAFT_BUILD_INFO=1
SNAPCRAFT_PROJECT_NAME=pivx
https_proxy=http://10.10.10.1:8222/
http_proxy=http://10.10.10.1:8222/
SNAP_CONTEXT=Mp2T2tRpD8LXh5Fd4BjFdJH4218jtpfGvGTdVPFhkk3W
SNAP_VERSION=3.4
PKG_CONFIG_PATH=:/build/pivx/parts/pivx/install/usr/share/pkgconfig
SNAP_INSTANCE_NAME=snapcraft
SNAPCRAFT_BUILD_ENVIRONMENT=host
PATH=/build/pivx/stage/usr/sbin:/build/pivx/stage/usr/bin:/build/pivx/stage/sbin:/build/pivx/stage/bin:/build/pivx/parts/pivx/install/usr/sbin:/build/pivx/parts/pivx/install/usr/bin:/build/pivx/parts/pivx/install/sbin:/build/pivx/parts/pivx/install/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin:/snap/snapcraft/2836/bin/scriptlet-bin
SNAPCRAFT_PRIME=/build/pivx/prime
SNAP_DATA=/var/snap/snapcraft/2836
XDG_RUNTIME_DIR=/run/user/0/snap.snapcraft
CFLAGS= -I/build/pivx/parts/pivx/install/usr/include
PERL5LIB=/build/pivx/stage/usr/share/perl5/
SNAPCRAFTCTL_FEEDBACK_FIFO=/tmp/tmp4k9vegxa/call_feedback
SNAPCRAFT_ARCH_TRIPLET=aarch64-linux-gnu
LANG=C.UTF-8
SNAPCRAFT_INTERPRETER=/snap/snapcraft/2836/usr/bin/python3
SNAPCRAFT_PART_BUILD=/build/pivx/parts/pivx/build
SNAP_USER_COMMON=/root/snap/snapcraft/common
GIT_PROXY_COMMAND=/usr/local/bin/snap-git-proxy
SNAP_ARCH=arm64
SNAP_COOKIE=Mp2T2tRpD8LXh5Fd4BjFdJH4218jtpfGvGTdVPFhkk3W
CPPFLAGS= -I/build/pivx/parts/pivx/install/usr/include
SHELL=/bin/sh
SNAP_REEXEC=
SNAP_NAME=snapcraft
SNAPCRAFTCTL_CALL_FIFO=/tmp/tmp4k9vegxa/function_call
PWD=/build/pivx/parts/pivx/build
SNAP=/snap/snapcraft/2836
SNAPCRAFT_PART_INSTALL=/build/pivx/parts/pivx/install
```

Icons which are used for the daemon in snap builds:
![grafik](https://raw.githubusercontent.com/cevap/PIVX/master/snap/local/desktop/pivxd.ico) ![grafik](https://raw.githubusercontent.com/cevap/PIVX/master/snap/local/desktop/pivxd_testnet.ico) ![grafik](https://raw.githubusercontent.com/cevap/PIVX/master/snap/local/desktop/pivxd_regtest.ico)

cevap commented:
updated to current master and squashed snap/ folder commits in one (https://github.com/PIVX-Project/PIVX/commit/90c0265377d48320c9f2f1cf16860284d47bb0ab)

Fuzzbawls commented:
@cevap thank you for your continued updates on this. I have been doing some of my own research as well, and I think that this has great potential.

One of the things I've been researching is the possibility to have the `snap` directory hosted in another repository and use github webhooks from **this** repo to signal that a new build should be done.

cevap commented:
you are welcome, just trying to pimp up ;) our sources build system, thank you guys for keeping pivx up to date (as far as it could be said in crypto).

For your question about hosting it somewhere else, yes, it is very easy. Simply create separate repo for snapcraft, place snap folder into it or (root of snap). On snapcraft/launchpad, you import then this repository, where it will build from defined source in snapcraft.yaml. So, yes, snap folder can be moved wherever you want, it would also work as a subdir on current repo.

I agree that snap is very good thing, its better than travis as it does really build and release builds. Beside that, snap's can have additional security measures and much much more. But for me the biggest advantage is if I keep it in my projects folder (for edge channel only), as then, I do not need to trigger manually builds, they are then built on each change on master (just like travis).

Snap offers not only much faster and better way than to release gitian, especially for projects where you dont have community which does gitian builds in the background, but it offers every maintainer/contributor simple ability to use/share/test/release their binaries, which is way simpler than gitian, only snap name and repo changed and it works, no need for Xpage guide and manual mistakes during that process.

Snap is also currently not replaceable for developers/testers, things like to have builds from each commit lets you always access those builds with ease, on any device. For me, this is the biggest example, especially when it comes to heavy testing in frequent projects where daily commits happen.

Combination with launchpad, I use it for now as beta channel which is also development one (not stable). It means, edge is built and released from snapcraft on each change on master (very unstable stable), launchpad does not build on each commit but around 5-6 times a day repo is updated and if changes exist, launchpad builds then from same commit SNAP and DEB packages which ensures me (devs) that everybody does use same binary (compiled from same hash), regardless if it is a deb package or snap.

Candidate and Releases, for them you could create separate repository because RC's and Releases (_candidate and stable channels, marked as non development, release channels_)  are to be built and released only manually (I think this is out of question if it should be automatic).

Pivx has debian folder in contrib, where you do have desktop files, manpages etc.., and that can be used by snap, I just did not patch it that way as I really have no clue about the end design which pivx will decide to have, but in general, I would simply add a patch which does patch those desktop.files and adaption/changes for snap are done on the fly. For icons it would be the same, it depends from how much scripting you want to have in your script, but you could create simply a mask with "snap build" or similar and patch existing images with a layer over existing image with "snap build", that way, you would not require any additional png's/icons except that one frame.

Current snap does build on each architecture for the same architecture. Gitian builds are all built on one platform for all other. This can be done by snapcraft too, I just did not apply it here, but yeah, it could be.

There are quite many cool things with snapcraft, as example, you could also include/install now offline wallet and any tools that you want or need. Check plugs out which I use, those are all compatible in strict environment which is important, as if we switch to classic, then pivx will not be available on devices which support only strict one (like RPi). I've added pulseaudio, not sure why, but I am quite sure pivx will at some time enable some sound for incoming transactions and then snap would require ability to access sound device, with that plug it should work out of the box now.

Above reasons/possibilities are just a small part of what is possible, you would need to read the docu to see what is possible. When I first time read it, I was confused by some thing until I tried them out as I expected it to be similar to deb packaging and was surprised at the end how much less complicated snap is. Not sure if launchpad still has restriction for number of builds for deb per day, for snapcraft there is no restriction.

This means, in fact, you can and should keep snapcraft in snap folder if you want snapcraft to build and release automaticaly over snapcraft or launchpad (on launchpad you probably can make some workaround, did look into it for now as its not required for me).

Pivx has several devs and whatever they develop, if they also have snap builds of whatever they do, then it makes it also much easier, but mainly time efficient, to test them. But as I said, there is quite no limits for creativity, I just tried to keep the script simple so that everybody can change/adapt it to its needs.

Fuzzbawls commented:
yeah my thought right now is to host the `snap`, `debian` and `rpm` directories all in one separate repository specifically used for packaging purposes. that would cut down on the number of non-source related commits needed here by quite a lot as standards change over time.

I did discover the fact that the `source:` line in the `.yaml` file can be used to point to a different repository, and i think that would be sufficient enough for "nightly" builds of the master branch. I also have seen some reference to a `{latest-tag}` logic check that may be of interest.

I don't see snaps full on replacing gitian, but rather as an alternative distribution means for those that do not absolutely require verification/validation of determinism and/or GPG key checking against multiple sources.

cevap commented:
yes, you are right. I did not create this config to compile from its own source (downloaded with snapcraft.yaml), but I use source defined in snapcraft.yaml, which does not require you to have sources where your snap is, on snapcraft server, when you test it, you simply push snapcraft folder with snapcraft.yaml, the rest is then done from the script. 

So, yes, it makes sense to keep them in one line especially for releases. If you check sources part, you will see that you can define also a branch, where I think, that releases should always be built from master, meaning, when marking release as true and releasing, then pivx creates anyway a tag, where I would put then in same commit changes from devel to stable and master branch to vX.X.X in snapcraft.yaml. That way, you actually do always build from each commit on master, where the one where the release happens will anyway build release version which does not require second build. 

All that means just, that if you actually keep it automatic, then you dont have to care about when to trigger, you only need to follow release notes process defined in your doc/ (_needs snap t be added btw,_). And that means, that with automatic builds, you will have also stable releases and release candidates. Remember, `grade: devel` can not be published to stable channels, `grade: stable` **can be** published to development channels. With that, it means for you that manually release process is then not to trigger the build, but simply to click on release of that specific bin on snapcraft store. Now if you want to automate also this, then your suggestion fits very well, as then you will probably have it on another repo where you will create new snap on launchpad and mark release channel stable/release. With that, when your launchpad builds debian packages, you would have snap builds from the same one and as it is a release and if you mark it as stable, then it will be published as stable.

However, I strongly do not recommend making stable release automatic! So, if you want to automate stable releases, lets assume that they are always candidates, with that, you simply go to your store page, and release candidate channel versions as release. That way:

1. you are sure that you make stable (final release) manually
2. idont forget, snap's update automaticly

cevap commented:
it is not replacement for gitian simply because it does not build mac and windows binaries, with that, gitian can't be turned off.

You are also correct about deterministic builds, but on the same time, it was required not only to verify for reasons of manipulation, but also if there would be a faulty build. For that reason snap's logic was actually to use shared libs (just like deb packages). But that 

1. causes issues for us, example is bdb which has to be patched on the fly to use snaps folders instead of systems
2. we can't define strictly version of dependencies, which is important.

Later, snap actually changed its path in saying that all builds should actually by possibility build static. With that it fits well as well as alternative to dynamic bins from deb packages.

However, for linux users or lets say masternode owners etc... . I doubt that run windows or mac servers and most people that I know have different coins which makes it quite hard for them to update their clients and it requires complicated scripting sometimes. With snap's you provide auto update system for those channels, so, if somebody's masternode stalled due to client upgrade which he is not aware of, then he would only need to restart hist client, as his system did upgrade it.

This is why I say, that manually released bins should be released only manually. About determinism, who said, that you cant build snap deterministic where you would simply print required files in log and push them to as example snap.sigs? It would be simply adding few lines to the snapcraft.yaml, so, to have deterministic builds is with SNAP actually **much** easier than with gitian, mainly much, much faster for your users. As gitian sigs require few people to build it, otherwise they are useless, same would be for snap, where actually most of users are able to setup snap in their git and thats it! No need to create VM's, wait for hours, learn how gitian works .... etc.

Determinism was on my to do list and I can guarantee you that snap builds can be released same way as gitian like I explained. For hardcoders, well, they can also build like gitian on local system and release snap directly from your pc. At the end of the day, gitian offers only signatures which need to be checked. How many people in PIVX do you know who do check it on each bin they run? On snap, you have not only snap.sigs, but you can always use/refresh and try them out. You clearly see the difference, the only missing things for snap for now are:

1. Snap's in MS Store
2. Snap's in Apple Store

Microsoft already not only releases officialy things on snap (like VSCode), but I bet with you, at some point, they will add snap to their system, as you can already now use Snap with Windows 10 and Unix shell. However, snapcraft does not offer for now windolls and apples arch's, point is, as soon as they do, you would be ready, as pivx can be cross compiled.

Fuzzbawls commented:
Noticing the TravisCI linter failures. if the two new `.patch` files cannot have their trailing whitespace removed, then they can be added as explicit exclusions to the whitespace linting script.

Another thought, I read somewhere that the `snap` directory could be placed under `build-aux` somewhere in the Snapcraft docs, I think that would be more suitable than directly under the tree's root, at least until it can be relocated to an external repository entirely (if that turns out to be a fully viable option).

I'd also like to thank you again for sticking with this for so long, and apologize for the excessive time it's taken to get this merged. Stick with it though! This is certainly a feature we'd like to add as it offers a more well-rounded user experience.

cevap commented:
@Fuzzbawls sorry for late reply, was on holidays

##### about build-aux/snap

yes, it can be placed into build-aux/snap, I can move it, nothing has to be changed for that, just moving the snap folder to build-aux/snap, [here](https://docs.snapcraft.io/release-notes-snapcraft-3-3) (ref https://github.com/PIVX-Project/PIVX/pull/782#issuecomment-493394570) is documentation part about it (supported since v3):

> #### Alternate directory for snapcraft.yaml
> 
> **build-aux/snap** is now supported as an alternative directory for _snapcraft.yaml_ and its assets (i.e.; hooks, gui, …).
>  
> To avoid confusion, snapcraft now display what directory it is picking for assets, depending on where the snapcraft.yaml is found. It will only pick **build-aux/snap** for assets if the **snapcraft.yaml** is found in that path.

##### about linter

ok, will whitelist them, however, after it can be removed as those patches will probably not change and if, probably no new empty lines will be added.

Could you check my first post here and see the checklist? I would like to to give pivx label to the official pivx account on snapcraft/launchpad, did you already create a user? If yes, write here simlply which user/email it is and I will transfer the ownership and I will make my builds private so that they dont appear in snap store/Software center, only pivx's official should be there to spare confusion.

##### some bugs non related to snapcraft.yaml

I got some devices for testing on other architectures, many linux OS's for those devices use quite old kernel and most people have snap2 preinstalled which is broken and does not work. In fact that is quite disappointing as the problem lays not in snap itself but in linux distribution, as example ubuntu server or debian 9 with rock4b, it uses kernel 4.4 and snap was broken resulting in no ability to use snap at all. If you get snapd from github and compile it, then it works on arm/arm64 with rockpi devices. 

do we have anybody in community with those other architectures like s390x? Would be good to test it before merging, as it is also good to know if those architectures have troubles with snapd.

After all that is finished, we should get back to the question of determinism and if there is at all requirement to upload it anywhere if it can be printed in build logs at the end of the log file, all bins can be downloaded manually and can be checked, but that cant be part of snapcraft.yaml because release can happen only after snapcraft.yaml is successfully finished. Any suggested changes to this?

cevap commented:
@Fuzzbawls updated sources to current master, added patches to whitelist, moved folder to build-aux/snap.

Snapcraft.io seems not to be in version 3.3 or higher and so it does not pick up automaticly build-aux (thats quite new), but I guess launchpad builds would if you set it to bionic.

here a confirmation that it works:
![grafik](https://user-images.githubusercontent.com/24996551/58453651-3df45180-811c-11e9-83cb-dfb73787450e.png)

cevap commented:
@Fuzzbawls @Mrs-X just as note, to build now `clean` has changed to `destructive-mode`, to build now, run `snapcraft --destructive-mode`. Do you want it to be moved back to ./snap for v3.3 release?

Fuzzbawls commented:
@cevap could you please join our Dev discord https://discord.gg/ra3atpD
